File: 数据结构/树/num199.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> res;
    vector<int> rightSideView(TreeNode* root) {
        dfs(root, 0);
        return res;
    }
    void dfs(TreeNode* root, int depth){
        if (root == NULL){return;}
        if (res.size() <= depth){res.push_back(-1);}
        res[depth] = root->val;
        dfs(root->left, depth+1);
        dfs(root->right, depth+1);
    }
};

File: 数据结构/树/num199.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def bfs(queue):
            while queue:
                nxt = []
                res.append(queue[-1].val)

                for node in queue:
                    if node.left: nxt.append(node.left)
                    if node.right: nxt.append(node.right)
                queue = nxt
        
        if not root:
            return []
            
        res = []
        bfs([root])
        return res
